File: ios/ExpoDigitalInkRecognitionModule.swift
import ExpoModulesCore
import MLKitDigitalInkRecognition

public class ExpoDigitalInkRecognitionModule: Module {
  public func definition() -> ModuleDefinition {

    Name("ExpoDigitalInkRecognition")

         AsyncFunction("downloadModel") { (languageTag: String) async throws in
            
            guard let identifier = DigitalInkRecognitionModelIdentifier(forLanguageTag: languageTag) else {
                throw NSError(domain: "DigitalInk", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid language tag"])
            }
            
            let model = DigitalInkRecognitionModel(modelIdentifier: identifier)
            let manager = ModelManager.modelManager()
            let conditions = ModelDownloadConditions(allowsCellularAccess: true, allowsBackgroundDownloading: true)
            
            // 1. Start the download (this no longer takes a closure)
            manager.download(model, conditions: conditions)
            
            // 2. Wait for the notification that the download finished
            _ = await NotificationCenter.default.notifications(named: .mlkitModelDownloadDidSucceed).first { notification in
                guard let downloadedModel = notification.userInfo?[ModelDownloadUserInfoKey.remoteModel.rawValue] as? DigitalInkRecognitionModel else { return false }
                return downloadedModel.modelIdentifier == identifier
            }
        }
        
        
        AsyncFunction("getDownloadedModels") { () -> [String] in
            let manager = ModelManager.modelManager()
            // Enumerate all available Digital Ink model identifiers and filter those already downloaded
            var downloaded: [String] = []
            let identifiers = DigitalInkRecognitionModelIdentifier.allModelIdentifiers()
            for identifier in identifiers {
                let model = DigitalInkRecognitionModel(modelIdentifier: identifier)
                if manager.isModelDownloaded(model) {
                    downloaded.append(identifier.languageTag)
                }
            }
            return downloaded
        }
        
        AsyncFunction("deleteModel") { (languageTag: String) async throws in
            guard let identifier = DigitalInkRecognitionModelIdentifier(
                forLanguageTag: languageTag
            ) else {
                throw NSError(
                    domain: "DigitalInk",
                    code: 1,
                    userInfo: [NSLocalizedDescriptionKey: "Invalid language tag"]
                )
            }
            
            let model = DigitalInkRecognitionModel(modelIdentifier: identifier)
            let manager = ModelManager.modelManager()
            try await manager.deleteDownloadedModel(model)
        }
        
        AsyncFunction("recognize") { (languageTag: String, strokes: [[[String: Double]]]) async throws -> [String: Any] in
            
            guard let identifier = DigitalInkRecognitionModelIdentifier(forLanguageTag: languageTag) else {
                throw NSError(domain: "DigitalInk", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid language tag"])
            }
            
            let model = DigitalInkRecognitionModel(modelIdentifier: identifier)
            
            // Safety check: Is the model actually there?
            if !ModelManager.modelManager().isModelDownloaded(model) {
                throw NSError(domain: "DigitalInk", code: 2, userInfo: [NSLocalizedDescriptionKey: "Model not downloaded"])
            }
            
            let options = DigitalInkRecognizerOptions(model: model)
            let recognizer = DigitalInkRecognizer.digitalInkRecognizer(options: options)
            
            // FIX: Single loop iteration for strokes
            var strokesArray: [Stroke] = []
            for strokeData in strokes {
                var points: [StrokePoint] = []
                for point in strokeData {
                    guard let x = point["x"], let y = point["y"], let t = point["t"] else { continue }
                    points.append(StrokePoint(x: Float(x), y: Float(y), t: Int(t)))
                }
                strokesArray.append(Stroke(points: points))
            }
            
            let ink = Ink(strokes: strokesArray)
            let result = try await recognizer.recognize(ink: ink)
            
            let candidates = result.candidates.map {
                ["text": $0.text, "score": $0.score?.doubleValue ?? 0.0]
            }
            
            return ["candidates": candidates]
        }
        
    }
}

File: android/src/main/java/expo/modules/digitalinkrecognition/ExpoDigitalInkRecognitionModule.kt
package expo.modules.digitalinkrecognition

import com.google.mlkit.vision.digitalink.*
import com.google.mlkit.common.model.RemoteModelManager
import com.google.mlkit.common.model.DownloadConditions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async


class ExpoDigitalInkRecognitionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoDigitalInkRecognition")

    // We use the CoroutineScope provided by the Module (appContext)
    AsyncFunction("downloadModel") Coroutine { languageTag: String ->
      val identifier = DigitalInkRecognitionModelIdentifier.fromLanguageTag(languageTag)
        ?: throw Exception("Invalid language tag")

      val model = DigitalInkRecognitionModel.builder(identifier).build()
      val manager = RemoteModelManager.getInstance()
      val conditions = DownloadConditions.Builder().build()

      // This ensures the Task is awaited within a coroutine context the compiler trusts
      manager.download(model, conditions).await()
      "success"
    }

    AsyncFunction("getDownloadedModels") Coroutine { ->
      val manager = RemoteModelManager.getInstance()
      val models = manager.getDownloadedModels(DigitalInkRecognitionModel::class.java).await()
      models.map { it.modelIdentifier?.languageTag ?: "unknown" }
    }

    AsyncFunction("recognize") Coroutine { languageTag: String, strokes: List<List<Map<String, Any>>> ->
      val identifier = DigitalInkRecognitionModelIdentifier.fromLanguageTag(languageTag)
        ?: throw Exception("Model not found")

      val model = DigitalInkRecognitionModel.builder(identifier).build()
      val recognizer = DigitalInkRecognition.getClient(
        DigitalInkRecognizerOptions.builder(model).build()
      )

      val inkBuilder = Ink.builder()
      for (stroke in strokes) {
        val strokeBuilder = Ink.Stroke.builder()
        for (point in stroke) {
          val x = (point["x"] as? Number)?.toFloat() ?: 0f
          val y = (point["y"] as? Number)?.toFloat() ?: 0f
          val t = (point["t"] as? Number)?.toLong() ?: 0L
          strokeBuilder.addPoint(Ink.Point.create(x, y, t))
        }
        inkBuilder.addStroke(strokeBuilder.build())
      }

      // Explicitly awaiting the Task result
      val result = recognizer.recognize(inkBuilder.build()).await()

      val candidates = result.candidates.map { candidate ->
        mapOf(
          "text" to candidate.text,
          "score" to (candidate.score ?: 0.0f)
        )
      }

      mapOf("candidates" to candidates)
    }
  }
}